framework/src/base/MooseServer.C

```diff
@@ -721,13 +743,177 @@ MooseServer::getEnumsAndDocs(MooseEnumType & moose_enum_param,
 }
 
 bool
-MooseServer::gatherDocumentDefinitionLocations(wasp::DataArray & /* definitionLocations */,
-                                               int /* line */,
-                                               int /* character */)
+MooseServer::gatherDocumentDefinitionLocations(wasp::DataArray & definitionLocations,
+                                               int line,
+                                               int character)
+{
+  Factory & factory = _check_app->getFactory();
+
+  // return without any definition locations added when parser root is null
+  if (!_check_app || !_check_app->parser()._root ||
+      _check_app->parser()._root->getNodeView().is_null())
+    return true;
+
+  // find hit node for zero based request line and column number from input
+  wasp::HITNodeView view_root = _check_app->parser()._root->getNodeView();
+  wasp::HITNodeView request_context =
+      wasp::findNodeUnderLineColumn(view_root, line + 1, character + 1);
+
+  // return without any definition locations added when node not value type
+  if (request_context.type() != wasp::VALUE)
+    return true;
+
+  // get name of parameter node parent of value and value string from input
+  std::string param_name = request_context.has_parent() ? request_context.parent().name() : "";
+  std::string val_string = request_context.last_as_string();
+
+  // add source code location if type parameter with registered object name
```

Comment:
Does this link object in input -> source file?

Reply:
@loganharbour -

Yes, this makes `type=ObjectName` work like goto-definition in @dschwen's VSCode plugin as he [requested](https://github.com/idaholab/moose/pull/25305#issuecomment-1749286453).

It returns the path to the `ObjectName.C` source code file where the object was registered for the GUI to use.

I talked to him previously about what to do when the source file does not exist as in executable-only installs.

And it was decided the best thing to do in that situation is send an empty response, so this check does that:

```cpp
    // return without any definition locations added if file cannot be read
    if (!file_line_info.isValid() ||
        !MooseUtils::checkFileReadable(file_line_info.file(), false, false, false))
```

This type --> source scenario is unit tested in the `DefinitionObjectTypeSource` case of `MooseServerTest`.